src/data/polynomial/ring_division.lean

```diff
@@ -227,16 +230,63 @@ variables [comm_semiring R] [no_zero_divisors R] {p q : R[X]}
-        (or.inr ∘ is_unit_of_mul_eq_one g _) (or.inl ∘ is_unit_of_mul_eq_one f _)⟩⟩,
+  refine ⟨λ h f g hf hg hp, (h.2 f g hp.symm).imp hf.eq_one_of_is_unit hg.eq_one_of_is_unit,
+    λ h, ⟨hp1 ∘ hp.eq_one_of_is_unit, λ f g hfg, (h (g * C f.leading_coeff) (f * C g.leading_coeff)
+      _ _ _).symm.imp (is_unit_of_mul_eq_one f _) (is_unit_of_mul_eq_one g _)⟩⟩,
   { rwa [monic, leading_coeff_mul, leading_coeff_C, ←leading_coeff_mul, mul_comm, ←hfg, ←monic] },
   { rwa [monic, leading_coeff_mul, leading_coeff_C, ←leading_coeff_mul, ←hfg, ←monic] },
   { rw [mul_mul_mul_comm, ←C_mul, ←leading_coeff_mul, ←hfg, hp.leading_coeff, C_1, mul_one,
         mul_comm, ←hfg] },
 end
 
+lemma monic.irreducible_iff_nat_degree (hp : p.monic) : irreducible p ↔
+  p ≠ 1 ∧ ∀ f g : R[X], f.monic → g.monic → f * g = p → f.nat_degree = 0 ∨ g.nat_degree = 0 :=
+begin
+  by_cases hp1 : p = 1, { simp [hp1] },
+  rw [irreducible_of_monic hp hp1, and_iff_right hp1],
+  refine forall₄_congr (λ a b ha hb, _),
+  rw [ha.nat_degree_eq_zero_iff_eq_one, hb.nat_degree_eq_zero_iff_eq_one],
+end
+
+lemma monic.irreducible_iff_nat_degree' (hp : p.monic) : irreducible p ↔ p ≠ 1 ∧
+  ∀ f g : R[X], f.monic → g.monic → f * g = p → g.nat_degree ∉ finset.Ioc 0 (p.nat_degree / 2) :=
+begin
+  simp_rw [hp.irreducible_iff_nat_degree,
+    finset.mem_Ioc, nat.le_div_iff_mul_le zero_lt_two, mul_two],
+  apply and_congr iff.rfl,
+  split; intros h f g hf hg he; subst he,
+  { refine λ ha, (h g f hg hf $ mul_comm g f).elim ha.1.ne' _,
+    rw hf.nat_degree_mul hg at ha,
+    exact (ha.1.trans_le $ (add_le_add_iff_right _).1 ha.2).ne' },
+  { simp_rw hf.nat_degree_mul hg at h,
+    by_contra hdp, simp_rw [not_or_distrib, ← ne.def, ← pos_iff_ne_zero] at hdp,
+    obtain hl|hl := le_total f.nat_degree g.nat_degree,
+    { exact h g f hg hf (mul_comm g f) ⟨hdp.1, add_le_add_left hl _⟩ },
+    { exact h f g hf hg rfl ⟨hdp.2, add_le_add_right hl _⟩ } },
+end
+
+lemma monic.not_irreducible_iff_exists_add_mul_eq_coeff (hm : p.monic) (hnd : p.nat_degree = 2) :
+  ¬ irreducible p ↔ ∃ c₁ c₂, p.coeff 0 = c₁ * c₂ ∧ p.coeff 1 = c₁ + c₂ :=
+begin
+  casesI subsingleton_or_nontrivial R,
+  { simpa only [nat_degree_of_subsingleton] using hnd },
+  rw [hm.irreducible_iff_nat_degree', hnd],
+  push_neg,
+  rw imp_iff_right,
+  split,
+  { rintros ⟨a, b, ha, hb, rfl, hdb|⟨⟨⟩⟩⟩,
+    have hda := hnd, rw [ha.nat_degree_mul hb, hdb] at hda,
+    use [a.coeff 0, b.coeff 0, by rw mul_coeff_zero],
+    convert ha.next_coeff_mul hb; simp only
+      [next_coeff, hnd, add_right_cancel hda, hdb, if_neg two_ne_zero, if_neg one_ne_zero] },
+  { rintros ⟨c₁, c₂, hmul, hadd⟩,
+    refine ⟨X + C c₁, X + C c₂, monic_X_add_C _, monic_X_add_C _, _, or.inl $ nat_degree_X_add_C _⟩,
+    rw [p.as_sum_range_C_mul_X_pow, hnd, finset.sum_range_succ, finset.sum_range_succ,
+      finset.sum_range_one, ← hnd, hm.coeff_nat_degree, hnd, hmul, hadd, C_mul, C_add, C_1],
+    ring },
+  { rintro rfl, simpa only [nat_degree_one] using hnd },
```

Comment:
With a couple short lemmas, you can prove this without all the extra monic irreducible lemmas.
```
lemma monic.eq_X_add_C (hm : p.monic) (hnd : p.nat_degree = 1) : p = X + C (p.coeff 0) :=
by rw [←one_mul X, ←C_1, ←hm.coeff_nat_degree, hnd, ←eq_X_add_C_of_nat_degree_le_one hnd.le]

lemma X_add_C_ne_one [nontrivial R] (r : R) : X + C r ≠ 1 :=
λ h, by simpa using congr_arg nat_degree h
```

```suggestion
  have hp1 : p ≠ 1 := λ hp1, by simpa only [hp1, nat_degree_one] using hnd,
  simp only [irreducible_of_monic hm hp1, not_forall, not_or_distrib],
  split,
  { rintros ⟨f, g, hf, hg, hfg, hf1, hg1⟩,
    have := congr_arg nat_degree hfg,
    simp [mt hf.nat_degree_eq_zero_iff_eq_one.mp hf1, mt hg.nat_degree_eq_zero_iff_eq_one.mp hg1,
      hnd, nat_degree_mul hf.ne_zero hg.ne_zero, nat.add_eq_two_iff] at this,
    rw [hf.eq_X_add_C this.1, hg.eq_X_add_C this.2, mul_add, add_mul, add_mul] at hfg,
    exact ⟨f.coeff 0, g.coeff 0, by simp [←hfg, coeff_C]⟩ },
  { rintros ⟨c₁, c₂, hmul, hadd⟩,
    refine ⟨_, _, monic_X_add_C c₁, monic_X_add_C c₂, _, X_add_C_ne_one c₁, X_add_C_ne_one c₂⟩,
    rw [p.as_sum_range_C_mul_X_pow, hnd, finset.sum_range_succ, finset.sum_range_succ,
      finset.sum_range_one, ← hnd, hm.coeff_nat_degree, hnd, hmul, hadd, C_mul, C_add, C_1],
    ring },
```

Reply:
Hi @tb65536 Thanks for the suggestion. Your lemmas look useful and I've added them to degree/definitions. Your proof is pretty short indeed but if I squeeze the two `simp`s it would grow by three lines. Moreover, the purpose of this PR is not to add a criterion just for quadratic polynomials, but also for monic polynomials in general. Therefore I think the current way of getting the final quadratic result in three step is reasonable. I think the `finset.Ioc` form (inspired by @Vierkantor's `deg f = 0 ∨ deg g = 0` form) will be the most useful, as you can easily break `g.nat_degree` into cases using `rintro/rcases` and get the corresponding `f.nat_degree` immediately using `add_right_cancel`, as demonstrated in the quadratic proof. I would be happy to delete the `=0 ∨ =0` form if you think it's unnecessary.

Reply:
Here's a shorter version with squeezed simps. I do think it's worth having a lemma like `monic.irreducible_iff_nat_degree'`, but I wonder if it might be nicer if it were more symmetric, with an interval for `f.nat_degree` and an interval for `g.nat_degree`. (And maybe it should use `set.Ioc` rather than `finset.Ioc`).
```suggestion
have p1 : p ≠ 1 := λ hp1, by simpa only [hp1, nat_degree_one] using hnd,
  simp only [irreducible_of_monic hm p1, not_forall, not_or_distrib],
  split,
  { rintros ⟨f, g, hf, hg, rfl, f1, g1⟩,
    simp only [hf.nat_degree_mul hg, nat.add_eq_two_iff, mt hf.nat_degree_eq_zero_iff_eq_one.mp f1,
      mt hg.nat_degree_eq_zero_iff_eq_one.mp g1, false_and, and_false, false_or, or_false] at hnd,
    refine ⟨f.coeff 0, g.coeff 0, mul_coeff_zero f g, _⟩,
    simpa only [next_coeff, hf.nat_degree_mul hg, hnd.1, hnd.2] using hf.next_coeff_mul hg },
  { rintros ⟨c₁, c₂, hmul, hadd⟩,
    refine ⟨_, _, monic_X_add_C c₁, monic_X_add_C c₂, _, X_add_C_ne_one c₁, X_add_C_ne_one c₂⟩,
    rw [p.as_sum_range_C_mul_X_pow, hnd, finset.sum_range_succ, finset.sum_range_succ,
      finset.sum_range_one, ← hnd, hm.coeff_nat_degree, hnd, hmul, hadd, C_mul, C_add, C_1],
    ring },
```

Reply:
Well, your proof has fewer lines but is 104 characters longer, but I've adopted your line 279-280, thanks!

If you want direct access to the two inequalities, then `set.Ioc` is better; but if you want to break down into cases, then `finset` gives you direct access to them (no need of invoking `fin_cases` to synthesize a `finset` from inequalities), which is more convenient here, e.g. in my proof I directly get `hdb : b.nat_degree = 0 + 1`.